Further improvement actions to rotate

hcq9102 opened the pull request:
Aim to further improve **_rotate_** algorithm
  -  Adding core partition functionality: enables the left and right sides of rotate to allocate the number of cores as needed
  - Removing the unneeded asynchronous code: decreases the overheads by removing the third step operation of rotate 

## Any background context you want to provide?
add num_cores facility #5763
Making sure num_cores is properly handled by parallel_executor #5792

hkaiser reviewed (dismissed):
LGTM, thanks!

hcq9102 commented on libs/core/algorithms/include/hpx/parallel/algorithms/rotate.hpp:
```diff
@@ -211,30 +214,67 @@ namespace hpx { namespace parallel { inline namespace v1 {
-            auto p = hpx::execution::parallel_task_policy()
-                         .on(policy.executor())
-                         .with(policy.parameters());
+            auto p = policy(hpx::execution::task);
```
error report:
/Users/runner/work/hpx/hpx/libs/core/algorithms/include/hpx/parallel/algorithms/rotate.hpp:250:22: error: no matching function for call to object of type 'hpx::execution::parallel_unsequenced_policy'
            auto p = policy(hpx::execution::task);

hkaiser replied:
I think this is because of https://github.com/STEllAR-GROUP/hpx/issues/5492. 

At a minimum, we need something like https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/executors/include/hpx/executors/execution_policy.hpp#L1324-L1327,` but just for the `task_policy_tag`. Most likely here as well, BTW:  https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/executors/include/hpx/executors/execution_policy.hpp#L1396-L1399

hcq9102 replied:
> I think this is because of #5492.
> 
> At a minimum, we need something like https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/executors/include/hpx/executors/execution_policy.hpp#L1324-L1327, but just for the `task_policy_tag`. Most likely here as well, BTW: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/executors/include/hpx/executors/execution_policy.hpp#L1396-L1399

Sorry for the late response, I got 2 drafts [draft_1](https://github.com/hcq9102/hpx/pull/1/files) & [draft_2](https://github.com/hcq9102/hpx/pull/2/files) (ignore the notes) for this issue, but I am not sure which one makes more sense. THANKS.

hkaiser commented on libs/core/algorithms/include/hpx/parallel/algorithms/rotate.hpp:
```diff
@@ -211,30 +215,67 @@ namespace hpx { namespace parallel { inline namespace v1 {
                 },
-                r.call2(p, non_seq(), first, new_first),
-                r.call2(p, non_seq(), new_first, last));
+                r.call2(execution::with_processing_units_count(p, numcores1),
```
What type is `p` of?

hkaiser replied:
There are three possible specializations for `with_processing_units_count_t` defined here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/executors/include/hpx/executors/execution_policy_parameters.hpp. In your case, I would expect for the second one to be viable. Apparently, one of the requires clauses prevents this from happening. If you look at the generated error message in more detail, you will most likely see it telling you why it decided not to instantiate the specialization of the `tag_invoke` overload.

hkaiser commented:
@hcq9102 I have added some missing support functions to the `unseq` policies. Things should be fine now.

hcq9102 commented:
> @hcq9102 I have added some missing support functions to the `unseq` policies. Things should be fine now. Please note, I have also rebased the branch onto master to avoid conflicts.

Great! Thanks ;)

hkaiser commented:
@hcq9102 FWIW, I think I have fixed the one remaining compilation problem.

hcq9102 commented:
> @hcq9102 FWIW, I think I have fixed the one remaining compilation problem.

Thanks a lot!

hkaiser reviewed (approved):
LGTM, thanks!

hkaiser commented:
bors merge

gonidelis commented:
@hcq9102  Amazing work!

hcq9102 commented:
> @hcq9102 Amazing work!

THANKS! ;)
